### ISSUE TITLE
Passing parameter to function/declaring variable with same name as type causes no-unused-vars error

### ISSUE THREAD

**Issue by bdsomer** [gajus/eslint-plugin-flowtype#341]
<details>
<summary>ESLint Config</summary>

```json
"eslintConfig": {
  "parser": "babel-eslint",
  "plugins": [
    "flowtype"
  ],
  "extends": [
    "eslint:recommended",
    "plugin:flowtype/recommended"
  ],
  "env": {
    "node": true,
    "es6": true,
    "mocha": true
  },
  "rules": {
    "semi": [
      "error",
      "always"
    ],
    "no-unused-vars": [
      "error",
      {
        "varsIgnorePattern": "__"
      }
    ],
    "curly": "error",
    "brace-style": "error",
    "multiline-comment-style": "error",
    "indent": [
      "error",
      2
    ],
    "space-before-blocks": "error",
    "space-before-function-paren": [
      "error",
      "never"
    ],
    "keyword-spacing": "error",
    "space-infix-ops": "error",
    "eol-last": 2,
    "quotes": [
      "error",
      "single"
    ],
    "no-eval": "error",
    "no-param-reassign": "error",
    "no-useless-escape": "warn"
  },
  "settings": {
    "onlyFilesWithFlowAnnotation": true
  }
}
```
</details>
<details>
<summary>Important dev-dependencies</summary>

```json
{
  "babel-cli": "6.26.0",
  "babel-core": "6.24.1",
  "babel-eslint": "8.2.5",
  "babel-plugin-transform-flow-strip-types": "6.22.0",
  "babel-plugin-transform-object-rest-spread": "6.26.0",
  "eslint": "5.0.0",
  "eslint-plugin-flowtype": "2.49.3"
}
```
</details>
<details>
<summary>Lint Command</summary>

```
./node_modules/.bin/eslint lib/testFile.js
```
</details>
Code being linted:

```
//@flow
import type {Auth} from './Auth';

module.exports = function(Auth: Auth) {
  require('../index.js')(Auth);
};
```
ER: No errors.
AR:
```
...../testFile.js
  2:14  error  'Auth' is defined but never used. Allowed unused vars must match /__/  no-unused-vars

✖ 1 problem (1 error, 0 warnings)
  2:14  error  'Auth' is defined but never used. Allowed unused vars must match /__/  no-unused-vars

✖ 1 problem (1 error, 0 warnings)
```

No errors occur when the parameter name is not `Auth`.

**Comment by gajus** [comment]
Is this something you could contribute a test case/ fix for?

**Comment by bdsomer** [comment]
@gajus not anymore. The bug randomly disappeared...

Should I still leave this open?

**Comment by gajus** [comment]
Got it. Will close it until there is a replicable case.

**Comment by bdsomer** [comment]
@gajus I'm actually going to reopen this, the way it actually disappeared is I switched `no-unused-vars` to `warn` level and was running `eslint` with the `--quiet` flag. I might be able to contribute something this weekend, depending on how busy I am.

**Comment by gajus** [comment]
Of course.

**Comment by bdsomer** [comment]
It also turns out that this is not only for function parameters. Defining variables will have the same effect. For example:

```javascript
//@flow
import type {Auth, AuthGenerator} from './Auth';

module.exports = function(AuthGeneratorInstance: AuthGenerator) {
  const Auth = AuthGeneratorInstance.createNewAuth();
};
```

ER: No errors.
AR:
```
...../testFile.js
  2:14  error  'Auth' is defined but never used. Allowed unused vars must match /__/  no-unused-vars

✖ 1 problem (1 error, 0 warnings)
  2:14  error  'Auth' is defined but never used. Allowed unused vars must match /__/  no-unused-vars

✖ 1 problem (1 error, 0 warnings)
```

**Comment by smvv** [comment]
The following patch triggers the error:
```
diff --git a/package.json b/package.json
index 13cf1f9..14d470f 100644
--- a/package.json
+++ b/package.json
@@ -11,13 +11,13 @@
   "devDependencies": {
     "ajv": "^6.5.0",
     "babel-cli": "^6.26.0",
-    "babel-eslint": "^6.1.2",
+    "babel-eslint": "^8.2.5",
     "babel-plugin-add-module-exports": "^0.2.1",
     "babel-plugin-transform-object-rest-spread": "^6.26.0",
     "babel-preset-env": "^1.7.0",
     "babel-register": "^6.26.0",
     "chai": "^4.1.2",
-    "eslint": "^3.16.0",
+    "eslint": "^4.19.0",
     "eslint-config-canonical": "1.8.1",
     "gitdown": "^2.5.2",
     "glob": "^7.1.2",
diff --git a/tests/rules/assertions/useFlowType.js b/tests/rules/assertions/useFlowType.js
index c6e0dd3..a2ea398 100644
--- a/tests/rules/assertions/useFlowType.js
+++ b/tests/rules/assertions/useFlowType.js
@@ -125,6 +125,7 @@ const ALWAYS_VALID = [
   '(x: A); type A = Y',
   'function x<A>(): A {}; x()',
   'import type A from "a"; (function(): A {})',
+  'import type {A} from "a"; (function(A: A) {A()})',
   '(function(): A {}); import type A from "a";',
   'declare interface A {}',
   'declare type A = {}'
```

Upgrading eslint and babel-eslint is important. Without the version upgrade, there is no error.

The output of `npm run test` is:
```
  3) no-unused-vars must not trigger an error in these cases
       valid
         import type {A} from "a"; (function(A: A) {A()}):

      AssertionError [ERR_ASSERTION]: Should have no errors but had 1: [ { ruleId: 'no-unused-vars must not trigger an error in these cases',
    severity: 1,
    message: '\'A\' is defined but never used.',
    line: 1,
    column: 14,
    nodeType: 'Identifier',
    source: 'import type {A} from "a"; (function(A: A) {A()})',
    endLine: 1,
    endColumn: 15 } ]
      + expected - actual

      -1
      +0

      at testValidTemplate (node_modules/eslint/lib/testers/rule-tester.js:418:20)
      at Context.RuleTester.it (node_modules/eslint/lib/testers/rule-tester.js:573:25)
```

FYI upgrading babel-eslint and eslint will also cause 12 errors without the added test case.

**Comment by gajus** [comment]
Closing as stale.
